runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt

```diff
@@ -0,0 +1,180 @@
+ * * There is no analogue to DynamicObjectParsers' useArrayPolymorphism flag
+ *
+ * @param encodeDefaults if false we will always write nulls explicitly.
+ *          If true we will omit properties which have a default value.
+ */
+class DynamicObjectSerializer(private val encodeDefaults: Boolean = false, val context: SerialModule = EmptyModule) {
+
+    fun <T> serialize(strategy: SerializationStrategy<T>, obj: T): dynamic {
+
+        // handle singular primitive and enum values immediately
+        if (strategy.descriptor.kind is PrimitiveKind) {
+            if (obj is Char) {
+                val value: Char = obj // without the explicit type annotation int value would be used
+                return value.toString()
+            }
+            if (obj is Long) {
+                return obj.toDouble()
+            }
+            return obj
+        }
+        if (strategy.descriptor.kind is UnionKind.ENUM_KIND) {
+            return strategy.descriptor.getElementName(obj.unsafeCast<Enum<*>>().ordinal)
+        }
+
+        val serializer = DynamicObjectEncoder(encodeDefaults)
+        serializer.encode(strategy, obj)
+        return serializer.result
+    }
+
+    @ImplicitReflectionSerializer
+    inline fun <reified T : Any> serialize(obj: T): dynamic = serialize(context.getContextualOrDefault(T::class), obj)
+
+    @ImplicitReflectionSerializer
+    inline fun <reified T : Any> serialize(obj: List<T?>): dynamic =
+        serialize(context.getContextualOrDefault(T::class).nullable.list, obj)
+}
+
+private class DynamicObjectEncoder(val encodeDefaults: Boolean) : ElementValueEncoder() {
+    private object NoOutputMark
+
+    var result: dynamic = NoOutputMark
+    lateinit var current: Node
+    var currentName: String? = null
+    var currentDescriptor: SerialDescriptor = StringDescriptor
+    var currentElementIsMapKey = false
+
+    class Node(val writeMode: WriteMode, val jsObject: dynamic) {
+        var index: Int = 0
+        lateinit var parent: Node
+    }
+
+    enum class WriteMode {
+        OBJ, MAP, LIST
+    }
+
+    override fun encodeElement(desc: SerialDescriptor, index: Int): Boolean {
+        current.index = index
+        currentDescriptor = desc
+
+        if (current.writeMode == WriteMode.MAP) {
+            currentElementIsMapKey = current.index % 2 == 0
+        } else {
+            currentName = desc.getElementName(index)
+        }
+        return true
+    }
+
+    override fun encodeValue(value: Any) {
+        if (currentElementIsMapKey) {
+            currentName = value.toString()
+        } else {
+            current.jsObject[currentName] = value
+        }
+    }
+
+    override fun encodeChar(value: Char) {
+        encodeValue(value.toString())
+    }
+
+    override fun encodeNull() {
+        if (currentElementIsMapKey) {
+            currentName = null
+        } else {
+            if (shouldEncodeElementDefault(currentDescriptor, current.index) &&
```

Comment:
If you want the converter to omit exactly `null`s, not all the default values (although I'm not sure if this is reliable in terms of parsing data back with `DynamicObjectParser`, then you can rename the flag and just check it there, without calling other methods.

Reply:
IMO `Dynamic` is not a serialisation format in the traditional sense, therefore I think it is fine to have options which will not play nice with the `DynamicObjectParser `. I will still set the default value of the option to encode null as `null`, so parsing the output of `DynamicObjectSerializer` should work by default.